File: src/research/counter.py
"""Counter-research — adversarial verification of claims.

Generates negation queries, finds alternative explanations,
and checks for replication failures. Mandatory for all research.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class CounterResult:
    """Result of counter-research on a claim."""
    original_claim: str
    negation_queries: list[str] = field(default_factory=list)
    counter_evidence: list[str] = field(default_factory=list)
    alternative_explanations: list[str] = field(default_factory=list)
    robustness: str = "moderate"  # robust, moderate, fragile, contested

    @property
    def is_contested(self) -> bool:
        return self.robustness == "contested"

    @property
    def has_counter_evidence(self) -> bool:
        return len(self.counter_evidence) > 0


def generate_negation_queries(claim: str) -> list[str]:
    """Generate search queries to find evidence against a claim.

    Strategies:
    1. Direct negation ("X is not true")
    2. Failure reports ("X failed", "X problems")
    3. Alternative framing ("instead of X", "better than X")
    4. Criticism ("X criticism", "X limitations")
    """
    queries = []

    # Direct negation
    queries.append(f"{claim} not true")
    queries.append(f"{claim} wrong")

    # Failure/problems
    queries.append(f"{claim} problems")
    queries.append(f"{claim} limitations")

    # Criticism
    queries.append(f"{claim} criticism")

    return queries


def assess_robustness(
    counter_count: int,
    alternative_count: int,
    supporting_count: int,
) -> str:
    """Assess claim robustness based on evidence balance.

    Returns: robust, moderate, fragile, or contested.
    """
    if supporting_count == 0:
        return "fragile"

    total_against = counter_count + alternative_count
    ratio = total_against / (total_against + supporting_count) if (total_against + supporting_count) > 0 else 0

    if ratio >= 0.5:
        return "contested"
    if ratio >= 0.3:
        return "fragile"
    if ratio >= 0.1:
        return "moderate"
    return "robust"


def run_counter_research(
    claim: str,
    counter_evidence: list[str] | None = None,
    alternative_explanations: list[str] | None = None,
    supporting_count: int = 1,
) -> CounterResult:
    """Run counter-research on a claim.

    Args:
        claim: The claim to verify adversarially
        counter_evidence: Evidence found against the claim
        alternative_explanations: Alternative explanations
        supporting_count: Number of supporting sources
    """
    counter = counter_evidence or []
    alternatives = alternative_explanations or []

    queries = generate_negation_queries(claim)
    robustness = assess_robustness(len(counter), len(alternatives), supporting_count)

    return CounterResult(
        original_claim=claim,
        negation_queries=queries,
        counter_evidence=counter,
        alternative_explanations=alternatives,
        robustness=robustness,
    )

File: rust/src/research/counter.rs
//! Counter-research — adversarial verification of claims.
//!
//! Generates negation queries, finds alternative explanations,
//! and checks for replication failures. Mandatory for all research.

use serde::{Deserialize, Serialize};

/// Result of counter-research on a claim.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CounterResult {
    pub original_claim: String,
    pub negation_queries: Vec<String>,
    pub counter_evidence: Vec<String>,
    pub alternative_explanations: Vec<String>,
    /// robust, moderate, fragile, contested
    pub robustness: String,
}

impl CounterResult {
    pub fn is_contested(&self) -> bool {
        self.robustness == "contested"
    }

    pub fn has_counter_evidence(&self) -> bool {
        !self.counter_evidence.is_empty()
    }
}

/// Generate search queries to find evidence against a claim.
///
/// Strategies: direct negation, failure reports, alternative framing, criticism.
pub fn generate_negation_queries(claim: &str) -> Vec<String> {
    vec![
        format!("{claim} not true"),
        format!("{claim} wrong"),
        format!("{claim} problems"),
        format!("{claim} limitations"),
        format!("{claim} criticism"),
    ]
}

/// Assess claim robustness based on evidence balance.
///
/// Returns: robust, moderate, fragile, or contested.
pub fn assess_robustness(counter_count: usize, alternative_count: usize, supporting_count: usize) -> &'static str {
    if supporting_count == 0 {
        return "fragile";
    }

    let total_against = counter_count + alternative_count;
    let total = total_against + supporting_count;
    let ratio = if total > 0 {
        total_against as f64 / total as f64
    } else {
        0.0
    };

    if ratio >= 0.5 {
        "contested"
    } else if ratio >= 0.3 {
        "fragile"
    } else if ratio >= 0.1 {
        "moderate"
    } else {
        "robust"
    }
}

/// Run counter-research on a claim.
pub fn run_counter_research(
    claim: &str,
    counter_evidence: Option<Vec<String>>,
    alternative_explanations: Option<Vec<String>>,
    supporting_count: usize,
) -> CounterResult {
    let counter = counter_evidence.unwrap_or_default();
    let alternatives = alternative_explanations.unwrap_or_default();
    let queries = generate_negation_queries(claim);
    let robustness = assess_robustness(counter.len(), alternatives.len(), supporting_count);

    CounterResult {
        original_claim: claim.to_string(),
        negation_queries: queries,
        counter_evidence: counter,
        alternative_explanations: alternatives,
        robustness: robustness.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_negation_queries() {
        let queries = generate_negation_queries("AI improves productivity");
        assert_eq!(queries.len(), 5);
        assert!(queries[0].contains("not true"));
        assert!(queries[4].contains("criticism"));
    }

    #[test]
    fn test_assess_robustness_robust() {
        assert_eq!(assess_robustness(0, 0, 10), "robust");
    }

    #[test]
    fn test_assess_robustness_moderate() {
        assert_eq!(assess_robustness(1, 0, 5), "moderate");
    }

    #[test]
    fn test_assess_robustness_fragile_no_support() {
        assert_eq!(assess_robustness(0, 0, 0), "fragile");
    }

    #[test]
    fn test_assess_robustness_fragile_high_ratio() {
        assert_eq!(assess_robustness(3, 0, 7), "fragile");
    }

    #[test]
    fn test_assess_robustness_contested() {
        assert_eq!(assess_robustness(5, 0, 5), "contested");
    }

    #[test]
    fn test_run_counter_research() {
        let result = run_counter_research(
            "Rust is faster than C",
            Some(vec!["Benchmark shows C wins in X scenario".to_string()]),
            None,
            3,
        );
        assert_eq!(result.original_claim, "Rust is faster than C");
        assert!(result.has_counter_evidence());
        assert_eq!(result.negation_queries.len(), 5);
    }

    #[test]
    fn test_run_counter_research_no_evidence() {
        let result = run_counter_research("some claim", None, None, 5);
        assert!(!result.has_counter_evidence());
        assert_eq!(result.robustness, "robust");
        assert!(!result.is_contested());
    }

    #[test]
    fn test_counter_result_contested() {
        let result = run_counter_research(
            "claim",
            Some(vec!["e1".to_string(), "e2".to_string(), "e3".to_string()]),
            Some(vec!["a1".to_string(), "a2".to_string()]),
            1,
        );
        assert!(result.is_contested());
    }
}
